### PR TITLE
Use configured errorgroupcallback

### PR DESCRIPTION
**Pull request by nrcventura**
Thank you for submitting a pull request.  Please review our [contributing guidelines](/CONTRIBUTING.md) and [code of conduct](https://opensource.newrelic.com/code-of-conduct/).

## Description

Updates the error group functionality mentioned in #1401. 
- Renames groupFingerprint to errorGroup
- Update error group attribute to not get generated if the error group is null, empty string, or just whitespace
- Gets the error group callback from configuration instead of the agent class

# Author Checklist
- [x] Unit tests, Integration tests, and Unbounded tests completed
- [ ] Performance testing completed with satisfactory results (if required)
- ~~[ ] [Agent Changelog](/src/Agent/CHANGELOG.md) or [Lambda Agent changelog](/src/AwsLambda/CHANGELOG.md) updated~~ to be updated in a separate PR 

# Reviewer Checklist
- [ ] Perform code review
- [ ] Pull request was adequately tested (new/existing tests, performance tests)
- [ ] Review Changelog

### CODE REVIEW

File: tests/Agent/UnitTests/Core.UnitTest/Errors/ErrorServiceTests.cs [hunk 1]
```diff
@@ -317,6 +317,79 @@ public void FromException_MarkErrorDataAsExpected_SameErrorClass_In_ExpectedClas
 
         }
 
+        [TestCase("test group")]
+        [TestCase(null)]
```
Comment (tippmar-nr) [inline tests/Agent/UnitTests/Core.UnitTest/Errors/ErrorServiceTests.cs:321]: Shouldn't `errorData.ErrorGroup` come back `null` for all three of these test cases?

**Review by tippmar-nr — APPROVED** [review]

File: tests/Agent/UnitTests/Core.UnitTest/Errors/ErrorServiceTests.cs (hunk 1 above)
Comment (nrcventura) [reply tests/Agent/UnitTests/Core.UnitTest/Errors/ErrorServiceTests.cs:321]: I'm keeping the abstraction very thin at this point and enforcing the data at the attribute level. So The AttributeDefinition has a conversion function to enforce that the invalid data is coerced to null in a single location. I considered putting it in the ErrorService, but it just seemed too easy to bypass the ErrorService and create an ErrorData instance directly. In fact a lot of the existing tests bypassed the ErrorService, which led to inconsistent behavior between tests.

**Comment by chynesNR** [comment]
Based on `AttributeDefinition.TrySetValue()`, it looks like we'll log whenever an attribute is set to null, even though it won't get sent because it's filtered out. If you don't have a callback set, I think we'll get a lot of log spam. Does that seem like the case?

**Comment by chynesNR** [comment]
Oh also, it doesn't have to be done in this PR, but the most recent version of the spec says we should be using the name `error.group.name` rather than `error_group` for the attribute.

**Comment by nrcventura** [comment]
> Based on `AttributeDefinition.TrySetValue()`, it looks like we'll log whenever an attribute is set to null, even though it won't get sent because it's filtered out. If you don't have a callback set, I think we'll get a lot of log spam. Does that seem like the case?

Based on the code in the HandleNullValue method and because the attribute is not a user attribute (it's an agent attribute) nothing should get logged. https://github.com/newrelic/newrelic-dotnet-agent/blob/90b8018fbb433aa554c24d133ad7df63147b648c/src/Agent/NewRelic/Agent/Core/Attributes/AttributeDefinition.cs#L354-L363.

**Comment by nrcventura** [comment]
> Oh also, it doesn't have to be done in this PR, but the most recent version of the spec says we should be using the name `error.group.name` rather than `error_group` for the attribute.

Thanks for the update. Yes, I'd like to handle the rename in a separate PR to make reviewing the changes easier.

**Comment by chynesNR** [comment]
> > Based on `AttributeDefinition.TrySetValue()`, it looks like we'll log whenever an attribute is set to null, even though it won't get sent because it's filtered out. If you don't have a callback set, I think we'll get a lot of log spam. Does that seem like the case?
> 
> Based on the code in the HandleNullValue method and because the attribute is not a user attribute (it's an agent attribute) nothing should get logged.
> 
> https://github.com/newrelic/newrelic-dotnet-agent/blob/90b8018fbb433aa554c24d133ad7df63147b648c/src/Agent/NewRelic/Agent/Core/Attributes/AttributeDefinition.cs#L354-L363

Whoops, my bad. I misread `!=` as `==`

**Review by chynesNR — APPROVED** [review]

**Pull request merged**